prov/lnx/include/lnx.h

```diff
@@ -0,0 +1,585 @@
+/*
+ * Copyright (c) 2022 ORNL. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef LNX_H
+#define LNX_H
+
+#define LNX_DEF_AV_SIZE 1024
+#define LNX_MAX_LOCAL_EPS 16
+#define LNX_IOV_LIMIT 5
+
+#define lnx_ep_rx_flags(lnx_ep) ((lnx_ep)->le_ep.rx_op_flags)
+
+struct local_prov_ep;
+
+struct lnx_match_attr {
+	fi_addr_t lm_addr;
+	uint64_t lm_tag;
+	uint64_t lm_ignore;
+	struct lnx_peer *lm_peer;
+	struct local_prov_ep *lm_cep;
+};
+
+/*
+ * . Each endpoint LNX manages will have an instance of this structure.
+ * . The structure has a pointer to the general shared cq.
+ *     . All events are written into that CQ.
+ * . The structure has a pointer to the core cq which the core provider
+ *   returns
+ * . The structure has an instance of the peer_cq which is unique for
+ *   handling communication with the constituent endpoint.
+ */
+struct lnx_peer_cq {
+	struct lnx_cq *lpc_shared_cq;
+	struct fid_peer_cq lpc_cq;
+	struct fid_cq *lpc_core_cq;
+};
+
+struct lnx_queue {
+	struct dlist_entry lq_queue;
+	dlist_func_t *lq_match_func;
+	ofi_spin_t lq_qlock;
```

Comment:
My 2 cents on this is locking should align with supported threading level. Typically, this results in coarse grain locking instead of fine grain per object locking. In addition, this allows for usage of ofi_genlock.